File: common/src/main/java/com/dreamdisplays/ytdlp/VideoMetadataCache.java
package com.dreamdisplays.ytdlp;

import me.inotsleep.utils.logging.LoggingManager;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@NullMarked
public final class VideoMetadataCache {

    private static final ConcurrentHashMap<String, YtVideoInfo> CACHE = new ConcurrentHashMap<>();
    private static final ConcurrentHashMap<String, Boolean> IN_FLIGHT = new ConcurrentHashMap<>();
    private static final ExecutorService EXEC = Executors.newSingleThreadExecutor(r -> {
        Thread t = new Thread(r, "DD-VideoMeta");
        t.setDaemon(true);
        return t;
    });

    private VideoMetadataCache() {
    }

    public static void put(String videoId, YtVideoInfo info) {
        if (videoId.isEmpty()) return;
        CACHE.put(videoId, info);
        // Also seed the title cache so the simpler consumers stay in sync
        VideoTitleCache.put(videoId, info.getTitle());
    }

    public static @Nullable YtVideoInfo get(String videoId) {
        return CACHE.get(videoId);
    }

    public static void requestAsync(String videoId) {
        if (videoId.isEmpty()) return;
        if (CACHE.containsKey(videoId)) return;
        if (IN_FLIGHT.putIfAbsent(videoId, Boolean.TRUE) != null) return;
        EXEC.submit(() -> fetchAndStore(videoId));
    }

    private static void fetchAndStore(String videoId) {
        try {
            YtVideoInfo meta = YouTubeWeb.metadata(videoId);
            if (meta != null) put(videoId, meta);
        } catch (Exception e) {
            LoggingManager.warn("Metadata fetch failed for " + videoId + ": " + e.getMessage());
        } finally {
            IN_FLIGHT.remove(videoId);
        }
    }
}

File: common/src/main/kotlin/com/dreamdisplays/ytdlp/VideoMetadataCache.kt
package com.dreamdisplays.ytdlp

import me.inotsleep.utils.logging.LoggingManager
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors

/**
 * Video metadata cache that stores `YtVideoInfo` objects in memory, keyed by video ID. This is used to avoid repeated
 * calls to `yt-dlp`
 */
object VideoMetadataCache {
    private val CACHE = ConcurrentHashMap<String, YtVideoInfo>()
    private val IN_FLIGHT = ConcurrentHashMap<String, Boolean>()
    private val EXEC = Executors.newSingleThreadExecutor { r ->
        Thread(r, "DD-VideoMeta").apply { isDaemon = true }
    }

    /** Stores [info] in the cache under [videoId] and also updates [VideoTitleCache]. */
    fun put(videoId: String, info: YtVideoInfo) {
        if (videoId.isEmpty()) return
        CACHE[videoId] = info
        VideoTitleCache.put(videoId, info.title)
    }

    /** Returns the cached [YtVideoInfo] for [videoId], or null if not yet fetched. */
    fun get(videoId: String): YtVideoInfo? = CACHE[videoId]

    /** Fetches and caches metadata for [videoId] in the background if it is not already cached or in flight. */
    fun requestAsync(videoId: String) {
        if (videoId.isEmpty()) return
        if (CACHE.containsKey(videoId)) return
        if (IN_FLIGHT.putIfAbsent(videoId, true) != null) return
        EXEC.submit { fetchAndStore(videoId) }
    }

    /** Calls [YouTubeInnerTube.metadata] for [videoId] and stores the result; logs a warning on failure. */
    private fun fetchAndStore(videoId: String) {
        try {
            YouTubeInnerTube.metadata(videoId)?.let { put(videoId, it) }
        } catch (e: Exception) {
            LoggingManager.warn("[VideoMetadataCache] Metadata fetch failed for $videoId: ${e.message}")
        } finally {
            IN_FLIGHT.remove(videoId)
        }
    }
}
